### DISCUSSION TITLE
Including C code cache problem?

### DISCUSSION THREAD

**Discussion by jorgeluismireles** [vlang/v#22157] [Questions and Answers]
I follow the https://docs.vlang.io/v-and-c.html#including-c-code doc to compile a c file I have. Then I had problems so copy, paste and update the code at https://github.com/vlang/v/tree/master/vlib/v/tests/project_with_c_code.

First I though was mandatory to have the C folder named exactly `c` and the c code as exactly `implementation.c` but I wasn't the case. 

Then notice like a **cache** is causing problems, since I modifed the `.c` file to get an expected assert error but the program run correctly as if the file wasn't touched. Even I changed the name of the file in order to cannot be found but no error was indicated for the new `v run`. These problems difficult the edition process.

Also something strange happens when I make a typo in the line `#flag @VMODROOT/mod1/code/libx.o` and I got this:
```
$ v run main.v
failed thirdparty object build cmd:
'c++' -std=gnu++17 -D_DEFAULT_SOURCE   -fwrapv  -fPIC -I "/home/jorge/project_with_c_code/mod1/code" 
-o '/home/jorge/.vmodules/cache/62/62adbcb43cb35192094e47fd20a554ff.module.mod1.o' 
-c '/home/jorge/project_with_c_code/mod1/code/libx.cpp'
builder error: c++: error: /home/jorge/project_with_c_code/mod1/code/libx.cpp: No such file or directory
c++: fatal error: no input files
compilation terminated.
```
Seems the compiler thinks we want to compile a c++ file.

Ideas? V 0.4.7 da228e9

**Comment by jorgeluismireles** [comment]
By the way, the folder https://github.com/vlang/v/tree/master/vlib/v/tests contains according to github more than 1000 entries, and not all are displayed by github and then we loose the chance to navigate there to look for more examples to try. Maybe making sub-folders can permit to see everything from github.

**Reply by spytheman** [reply]
I think that is now fixed - most of the tests there are now organized into subfolders.

**Comment by jorgeluismireles** [comment]
I have learn several things... I have two C libraries in a single app: One for the [serial port](https://github.com/sigrokproject/libserialport) and another for [can-utils](https://github.com/linux-can/can-utils). The later requires to prevent calling standar `linux/can.h` before the **can-utils** updated version and can be done with something like this:
### vcan/socketcan.c.v
```
#flag -I @VMODROOT/vcan/can-utils/
#flag -I @VMODROOT/vcan/can-utils/include/
#flag -I @VMODROOT/vcan/can-utils/include/linux/
#flag -I @VMODROOT/vcan/can-utils/include/linux/can/

#include <net/if.h>
#include <sys/epoll.h>
#include <sys/socket.h>
#include <linux/can.h>
#include <linux/can/error.h>
#include <lib.h>

#flag @VMODROOT/vcan/can-utils/lib.o
```
So far the app worked. But then I added to the app the serial port with my newbie unnecessary line `flag -I /usr/include`:
###  libserial/libserial.c.v
```
#pkgconfig --libs --cflags libserialport
#include <libserialport.h>
//#flag -I /usr/include // THIS caused problems to vcan/socketcan.c.v
#flag -l serialport
```
I didn't expect both C libs declarations would be mixed and the later affecting the former only when combined. I wonder if apps with more than two C libraries cause problems more difficult to solve. How to know the order of includes will be called by the compiler?

### `-show-c-output`
This flag gives a lot of details (including the location of the warnings in c.v code) but only when something is fatal, like when a wrong h file was added first and a definition is not found. But when the compiler only find warnings the details are not displayed and then is difficult to know what parts of c.v code to fix.

**Reply by spytheman** [reply]
Use `-showcc` .

**Reply by spytheman** [reply]
Another option, that is more verbose, but perhaps easier to read, since it shows each flag on its own line is to use:
`v -dump-c-flags - run program.v`
For example:
```
#0 07:10:00 ^ master /space/v/oo>v -dump-c-flags - examples/hello_world.v
-fwrapv
-o "/space/v/oo/examples/hello_world"
-D GC_THREADS=1
-D GC_BUILTIN_ATOMIC=1
-I "/home/delian/code/v/thirdparty/libgc/include"
"/tmp/v_1000/hello_world.01JEQFJT8Z5FJ3WWJ7KH5AV3PA.tmp.c"
-std=gnu99
-D_DEFAULT_SOURCE
-bt25
"/home/delian/code/v/thirdparty/tcc/lib/libgc.a"
-ldl
-lpthread
#0 07:10:01 ^ master /space/v/oo>
```

**Reply by spytheman** [reply]
You can also use `-dump-c-flags file.txt`, then the options will be stored in that file, and you can use an editor with change detection like notepad++, kate or emacs to open it, while you tweak the .v source.

**Reply by jorgeluismireles** [reply]
> You can also use `-dump-c-flags file.txt`.

Thanks. There were several questions in this discussion. But this flags should help detecting what directives are in use and track duplicated or omitted ones.

**Comment by spytheman** [comment]
As for the c++ compiler, I currently have no idea why it decided to use that, instead of a c compiler.

According to https://github.com/vlang/v/blob/master/vlib/v/builder/cc.v#L1082 , it will try to see if there is a corresponding `.c` file for each `#flag file.o`, and if it can not find a `.c` file, it will try to search for a corresponding `.cpp` file.

If your file was named with a .c extension, it should have used a c compiler 🤔 .
